secant/Features/Scan/UIKitBridge/ScanUIView.swift

```diff
@@ -0,0 +1,124 @@
+            doInitialSetup()
+        }
+    }
+
+    var onQRScanningDidFail: (() -> Void)?
+    var onQRScanningSucceededWithCode: ((String?) -> Void)?
+    var onQRScanningDidStop: (() -> Void)?
+    
+    required init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+    }
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+    }
+    
+    override public class var layerClass: AnyClass {
+        return AVCaptureVideoPreviewLayer.self
+    }
+    
+    override public var layer: AVCaptureVideoPreviewLayer {
+        return super.layer as? AVCaptureVideoPreviewLayer ?? AVCaptureVideoPreviewLayer()
+    }
+}
+
+extension ScanUIView {
+    func stopScanning() {
+        captureSession?.stopRunning()
+        onQRScanningDidStop?()
+    }
+    
+    private func doInitialSetup() {
+        clipsToBounds = true
+        captureSession = AVCaptureSession()
+        
+        guard let videoCaptureDevice = AVCaptureDevice.default(for: .video) else {
+            return
+        }
+        
+        let videoInput: AVCaptureDeviceInput
+        do {
+            videoInput = try AVCaptureDeviceInput(device: videoCaptureDevice)
+        } catch {
+            return
+        }
+        
+        if captureSession?.canAddInput(videoInput) ?? false {
+            captureSession?.addInput(videoInput)
+        } else {
+            scanningDidFail()
+            return
+        }
+        
+        let metadataOutput = AVCaptureMetadataOutput()
+        
+        if captureSession?.canAddOutput(metadataOutput) ?? false {
+            captureSession?.addOutput(metadataOutput)
+            
+            metadataOutput.setMetadataObjectsDelegate(self, queue: DispatchQueue.main)
+            metadataOutput.metadataObjectTypes = [.qr]
+        } else {
+            scanningDidFail()
+            return
+        }
+        
+        self.layer.session = captureSession
+        self.layer.videoGravity = .resizeAspectFill
+        
+        captureSession?.commitConfiguration()
+        captureSession?.startRunning()
+        metadataOutput.rectOfInterest = internalRectOfInterest
+    }
+    
+    func scanningDidFail() {
+        onQRScanningDidFail?()
+        captureSession = nil
+    }
+    
+    func found(code: String) {
+        onQRScanningSucceededWithCode?(code)
+    }
+}
+
+extension ScanUIView: AVCaptureMetadataOutputObjectsDelegate {
+    public func metadataOutput(
+        _ output: AVCaptureMetadataOutput,
+        didOutput metadataObjects: [AVMetadataObject],
+        from connection: AVCaptureConnection
+    ) {
+        if let metadataObject = metadataObjects.first {
+            guard let readableObject = metadataObject as? AVMetadataMachineReadableCodeObject else { return }
+            guard let stringValue = readableObject.stringValue else { return }
+            if !continueVideoSessionAfterScanning {
+                AudioServicesPlaySystemSound(SystemSoundID(kSystemSoundID_Vibrate))
```

Comment:
I think feedback could be delegated to the invoker if the ScanView.

Reply:
This is similar like the DispatchQuque... We "can't" pass any environment's data to the view but I'll think about this whole camera view again in a few days. A lot of instances are hardcoded (CaptureSession, Metadata, dispatch queue, feedback sound, ...). Ugly way would be to try to create equatable environment, pass it to the view as a closure (which would be passed every time the view re-renders) - that is very hacky solution.